Client generation with required sanitization

rahul1995 opened the pull request:
## Purpose
To create the OpenAPI spec, sanitize it to make compatible with Ballerina OpenAPI tool, and generate Client using it.

Issue Title: Introduce a Ballerina connector for Dropbox REST API
Issue link: https://github.com/ballerina-platform/ballerina-library/issues/7054

## Examples

## Checklist
- [x] Linked to an issue
- [ ] Updated the changelog
- [ ] Added tests
- [x] Updated the spec
- [ ] Checked native-image compatibility

rahul1995 committed:
Create OpenAPI yaml from the Postman collection at https://www.postman.com/dropbox-api/dropbox-s-public-workspace/collection/q1zdtug/dropbox-api-reference

rahul1995 committed:
Sanitize OpenAPI spec by adding content type and schema to every content

rahul1995 committed:
Generate client with required sanitization from OpenAPI spec

rahul1995 committed:
Update docs/spec/sanitations.md

Co-authored-by: Sumudu Nissanka <lnash94@users.noreply.github.com>

rahul1995 committed:
Add icon, license, and create reusable ref in openapi yaml for requests with empty body

rahul1995 committed:
Update the ballerina command used in the Sanitation doc

rahul1995 committed:
Use OAS JSON generated using https://blog.postman.com/creating-an-openapi-definition-from-a-collection-with-the-postman-api/ and removing the OAS YAML generated using https://p2o.defcon007.com/, and re-generated the client

rahul1995 committed:
Remove API version from endpoint and add it to the server url instead to centralize the versioning to the base URL

Nuvindu commented on ballerina/Ballerina.toml:
```diff
@@ -6,7 +6,7 @@ version = "0.1.0"
 license = ["Apache-2.0"]
 authors = ["Ballerina"]
 keywords = []
```
Shall we add some keywords here?
There is already an older connector for Dropbox. You can check for the keywords there. And may add some others if necessary
https://central.ballerina.io/ballerinax/dropbox/1.5.1

rahul1995 replied:
Done

rahul1995 committed:
Add keywords, copied from https://central.ballerina.io/ballerinax/dropbox/1.5.1

lnash94 commented:
@rahul1995, This generated OpenAPI looks good 😃, Here I have noticed a few improvements we can add. 

- Given that the OpenAPI specification has a lot of inline schemas for particular request bodies and responses, shall we move them to named schemas under the components section, define names that properly align with [Ballerina record naming best practices](https://learn-ballerina.github.io/best_practices/naming_conventions.html#:~:text=MAX_SIZE%20%3D%201000%3B-,Objects%20and%20records,-All%20object%20and), and refer to them in the relevant places?

**ex: some sample for the inline schema**

```json
"requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "type": "object",
                "properties": {
                  "photo": {
                    "type": "object",
                    "properties": {
                      ".tag": {
                        "type": "string",
                        "example": "base64_data"
                      },
                      "base64_data": {
                        "type": "string",
                        "example": "SW1hZ2UgZGF0YSBpbiBiYXNlNjQtZW5jb2RlZCBieXRlcy4gTm90IGEgdmFsaWQgZXhhbXBsZS4="
                      }
                    }
                  }
                }
              },
```
**After modified ,** 
the inline schema need to be sent under the component.schemas section with named object and reference it into relevant requestBody or response in operation.
```json
"component": {
          "schemas": {
            "ProfilePhoto": {
              "schema": {
                "type": "object",
                "properties": {
                  "photo": {
                    "type": "object",
                    "properties": {
                      ".tag": {
                        "type": "string",
                        "example": "base64_data"
                      },
                      "base64_data": {
                        "type": "string",
                        "example": "SW1hZ2UgZGF0YSBpbiBiYXNlNjQtZW5jb2RlZCBieXRlcy4gTm90IGEgdmFsaWQgZXhhbXBsZS4="
                      }
                    }
                  }
                }
              },
...
"requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "$ref": "#/component/schemas/ProfilePhoto"
                    }
                  }
                }
              },

lnash94 commented on ballerina/client.bal:
```diff
@@ -13,3 +16,2323 @@
+    #
+    # + headers - Headers to be sent with the request 
+    # + return - update response 
+    resource isolated function post file_requests/update(file_requests_update_body payload, UpdateHeaders headers = {}) returns inline_response_200_8|error {
```
These naming `file_requests_update_body`, `inline_response_200_8`  can be addressed to align with Ballerina Naming convention via this solution https://github.com/ballerina-platform/module-ballerinax-dropbox/pull/3#issuecomment-2440562817

rahul1995 commented:
@lnash94 Thanks for the review. I agree with your comment, but this is like 35k lines of json file. It's looking infeasible for me to finish it as an individual. Can I improve some main API endpoints for now and take this as a task for later (distributed among other folks if possible)?

rahul1995 committed:
Define `dropbox-api-path-root`, `dropbox-api-select-user`, and `dropbox-api-select-admin` parameters and reuse across API endpoints

lnash94 commented:
> @lnash94 Thanks for the review. I agree with your comment, but this is like 35k lines of json file. It's looking infeasible for me to finish it as an individual. Can I improve some main API endpoints for now and take this as a task for later (distributed among other folks if possible)?

Hi @rahul1995,

Yeah, it is understandable that this has a lot of endpoints. Until we provide the set of operations, I can bring some workarounds to you to address the pain points here.
**Workaround:**
1. Instead of the JSON format, let's use YAML format to increase readability and reduce unnecessary line counts. You can use this Swagger editor: https://editor-next.swagger.io/ to convert it into YAML format.
2. Then, you can use the same editor to generate a server or client using OpenAPI as the language. It will generate the YAML and automatically move inline objects into named records (you can use any preferred method here).
3. Once you receive the YAML, you can edit the name references by giving them meaningful names that align with Ballerina best practices. ex: `inline_response_200_73_event_category` **->** `EventCategoryResponse`

Please Note: during every modification lets check the modified details in the API is align with the original API documentation 😊.
Hope this workaround helps alleviate the issues you're experiencing. If you have any questions or need additional assistance, please don't hesitate to reach out to us. We're here to help 🙌 !

rahul1995 committed:
Use named schemes in OpenAPI spec

rahul1995 commented:
@lnash94 Thanks for sharing this method. It will save a lot of time.
I see that there are ~500 named schemas which I need to rename because many names are like `inline_response_xxx`, and others are not following the best practices.

Since these many will take a while, I have created a commit with a few renamed so that I get an early review on this.
Could you please review?

rahul1995 committed:
Prefer named schemas over inline schemas

lnash94 commented:
> @lnash94 Thanks for sharing this method. It will save a lot of time. I see that there are ~500 named schemas which I need to rename because many names are like `inline_response_xxx`, and others are not following the best practices.
> 
> Since these many will take a while, I have created a commit with a few renamed so that I get an early review on this. Could you please review?

Yeah sure , I will review :)

lnash94 commented on docs/spec/openapi.yaml:
```diff
@@ -0,0 +1,16368 @@
+                      ".tag": owner
+                    is_inside_team_folder: false
+                    is_team_folder: false
+                    link_me
```
```suggestion
    ProfilePhotoRequest:
```

lnash94 replied:
SetProfilePhotoRequest ->  ProfilePhotoRequest

lnash94 replied:
Please refer to this comment : https://github.com/ballerina-platform/module-ballerinax-dropbox/pull/3#issuecomment-2443138122

lnash94 commented on docs/spec/openapi.yaml:
```diff
@@ -0,0 +1,16368 @@
+                      ".tag": owner
+                    is_inside_team_folder: false
+                    is_team_folder: false
+                    link_me
```
```suggestion
    ProfilePhotoResponse:
```

lnash94 replied:
SetProfilePhotoResponse -> ProfilePhotoResponse

lnash94 commented on docs/spec/openapi.yaml:
```diff
@@ -0,0 +1,16368 @@
+                      ".tag": owner
+                    is_inside_team_folder: false
+                    is_team_folder: false
+                    link_me
```
```suggestion
    UserRequest:
```

lnash94 replied:
CheckUserRequest -> UserRequest

lnash94 commented on docs/spec/openapi.yaml:
```diff
@@ -0,0 +1,16368 @@
+                      ".tag": owner
+                    is_inside_team_folder: false
+                    is_team_folder: false
+                    link_me
```
```suggestion
    ManualContactsBatchRequest:
```

lnash94 replied:
DeleteManualContactsBatchRequest -> ManualContactsBatchRequest

lnash94 commented on docs/spec/openapi.yaml:
```diff
@@ -0,0 +1,16368 @@
+                      ".tag": owner
+                    is_inside_team_folder: false
+                    is_team_folder: false
+                    link_me
```
```suggestion
    PropertyRemovalDetails:
```

lnash94 replied:
RemovePropertiesRequest -> PropertyRemovalDetails

lnash94 commented on docs/spec/openapi.yaml:
```diff
@@ -0,0 +1,16368 @@
+                      ".tag": owner
+                    is_inside_team_folder: false
+                    is_team_folder: false
+                    link_me
```
it’s generally recommended to avoid using verbs as prefixes in record names. The best practice is to focus on the **state or data** the record represents rather than the action performed on it.
Shall we simplify the object name by minimizing action verb for clarity and Ballerina naming alignment?

ex: AddPropertiesRequest  -> NewPropertiesRequest, PropertiesRequest

lnash94 replied:
Please check and apply this to other relevant places.

lnash94 commented:
@rahul1995, avoiding using verbs as prefixes in record names is generally recommended. The best practice is to focus on the **state or data** the record represents rather than the action performed on it.

Shall we simplify the object name by minimizing the action verb for clarity and alignment with Ballerina's name?

ex: AddPropertiesRequest -> NewPropertiesRequest, PropertiesRequest

Please check and apply this to other relevant places.

rahul1995 commented:
@lnash94 I have made it like this only, it is that sometimes a particular API endpoint requires a specific format of request, that is where I am also creating records specific to that request body and naming it with verb. For example, I first created a record/schema `Photo` and then I created `SetProfilePhotoRequest` using `Photo` schema to define the format `SetProfilePhoto` API expects.

I have used the same practices as done in https://github.com/ballerina-platform/module-ballerinax-twitter/blob/main/docs/spec/openapi.json

rahul1995 committed:
Clean up

rahul1995 commented:
Hey @lnash94 thanks for the review :) I cleaned up a little bit and added PRs for [tests](https://github.com/ballerina-platform/module-ballerinax-dropbox/pull/4), [examples](https://github.com/ballerina-platform/module-ballerinax-dropbox/pull/6), and [documentations](https://github.com/ballerina-platform/module-ballerinax-dropbox/pull/5).

For the Schema naming, the current strategy is also in line with the Dropbox's official documentation, i.e. Dropbox also uses same naming conventions (`verb` + `API Name` + `Arg/Result`). For example, `AddTemplateResult` as shown in the screenshot attached for the [/templates/add_for_user API](https://www.dropbox.com/developers/documentation/http/documentation#file_properties-templates-add_for_user).

Please let me know how we should proceed further. 😊
<img width="1513" alt="Screenshot 2024-10-30 at 12 27 59 AM" src="https://github.com/user-attachments/assets/426e1825-47d4-4c9f-99ea-9802ec31537a">

lnash94 commented:
Hi @rahul1995 


> Hey @lnash94 thanks for the review :) I cleaned up a little bit and added PRs for [tests](https://github.com/ballerina-platform/module-ballerinax-dropbox/pull/4), [examples](https://github.com/ballerina-platform/module-ballerinax-dropbox/pull/6), and [documentations](https://github.com/ballerina-platform/module-ballerinax-dropbox/pull/5).
> 
> For the Schema naming, the current strategy is also in line with the Dropbox's official documentation, i.e. Dropbox also uses same naming conventions (`verb` + `API Name` + `Arg/Result`). For example, `AddTemplateResult` as shown in the screenshot attached for the [/templates/add_for_user API](https://www.dropbox.com/developers/documentation/http/documentation#file_properties-templates-add_for_user).
> 
> Please let me know how we should proceed further. 😊 <img alt="Screenshot 2024-10-30 at 12 27 59 AM" width="1513" src="https://private-user-images.githubusercontent.com/7647789/381274069-426e1825-47d4-4c9f-99ea-9802ec31537a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAyNjEzMDEsIm5iZiI6MTczMDI2MTAwMSwicGF0aCI6Ii83NjQ3Nzg5LzM4MTI3NDA2OS00MjZlMTgyNS00N2Q0LTRjOWYtOTllYS05ODAyZWMzMTUzN2EucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAzMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMzBUMDQwMzIxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MjE1MDlmNDY0OTdmYTg3MmY5ZDZhOGRhMWI3MzJlZjVjMGY5MTY1MjM5MGI1MmJkNjg5YzAxNDNkZjI1MzFlZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.0Rl-ZIvux2CZfBH6JeZ19Mmzfpl1atvS3_ik-102tQ0">

Let's proceed with the process (`verb` + `API Name` + `Arg/Result`) , it is fine :) .  lets follow the same thing which has been done in Dropbox.

rahul1995 committed:
Name schemas in same fashion as Dropbox

rahul1995 commented:
Thanks @lnash94. I have exactly matched these names as given in the documentation. Please review.

Also, I am open to contribute further outside of Hacktoberfest if needed :)
Learned a lot about OpenAPI during this task.

lnash94 commented on docs/spec/sanitations.md:
````diff
@@ -1,24 +1,37 @@
 
 ```bash
-# TODO: Add OpenAPI CLI command used to generate the client
+bal openapi -i docs/spec/openapi.json --mode client  --license docs/license.txt -o ballerina
````
```suggestion
bal openapi -i docs/spec/openapi.yaml --mode client  --license docs/license.txt -o ballerina
```

rahul1995 replied:
Thanks for pointing this out. I moved from yaml to json and updated it but forgot to change after moving back from json to yaml 😆

rahul1995 committed:
Update docs/spec/sanitations.md

Co-authored-by: Sumudu Nissanka <lnash94@users.noreply.github.com>

NipunaRanasinghe reviewed (approved):
LGTM
